File: 2024/day2/Python/day2.py
# -*- coding: utf-8 -*-
"""
@author: jkv
"""

def is_save(level):
    """
    Check if a level is safe based on given rules:
    - Consecutive differences must remain consistent in direction.
    - Absolute differences between elements must be between 1 and 3.
    """
    prev_diff = 0
    for i in range(len(level) - 1):
        diff = level[i + 1] - level[i]
        if prev_diff * diff < 0 or not (0 < abs(diff) <= 3):
            return False
        prev_diff = diff
    return True


def parse_input(file_path):
    """
    Parse input from a text file where each line represents a level.
    """
    with open(file_path, 'r') as f:
        return [[int(n) for n in line.split()] for line in f.readlines()]


def part_a(levels):
    """
    Solve Part A: Count how many levels are safe.
    """
    return sum(1 for level in levels if is_save(level))


def part_b(levels):
    """
    Solve Part B: Count how many levels can be made safe by removing one element.
    """
    safe_count = 0
    for level in levels:
        if is_save(level):
            safe_count += 1
            continue

        # Check if removing one element makes the level safe
        for i in range(len(level)):
            modified_level = level[:i] + level[i + 1:]
            if is_save(modified_level):
                safe_count += 1
                break

    return safe_count


if __name__ == "__main__":
    # Load test and input data
    test_file = "day2_test.txt"
    input_file = "day2_input.txt"

    test_levels = parse_input(test_file)
    input_levels = parse_input(input_file)

    # Run assertions for test data
    assert part_a(test_levels) == 2, "Part A Test failed!"
    assert part_b(test_levels) == 4, "Part B Test failed!"

    # Solve and print results for input data
    print("Part A:", part_a(input_levels))
    print("Part B:", part_b(input_levels))

File: 2024/day2/R/day2.R
# Function to calculate the sum of sequences satisfying a "safe" condition
safe_sequence_count <- function(file_path = "day2_input.txt") {
  # Read the input file line by line
  input_lines <- readLines(file_path)
  
  # Convert each line into a list of integers
  sequences <- sapply(strsplit(input_lines, " "), as.integer, simplify = FALSE)
  
  # Define the "safe" condition for a sequence
  is_safe <- function(sequence) {
    # Compute differences between adjacent elements
    differences <- diff(sequence)
    # Check if all differences are within a valid range
    all(differences > 0 & differences < 4) || all(differences < 0 & differences > -4)
  }
  
  # Apply the "safe" condition to each sequence and sum up the results
  safe_count <- sum(sapply(sequences, is_safe))
  
  return(safe_count)
}

#' Function to calculate the sum of sequences where any sub-sequence is "safe"
#' @param file_path Path to the input file
#' @export
safe_subsequence_count <- function(file_path = "day2_input.txt") {
  # Read the input file line by line
  input_lines <- readLines(file_path)
  
  # Convert each line into a list of integers
  sequences <- sapply(strsplit(input_lines, " "), as.integer, simplify = FALSE)
  
  # Define the "safe" condition for sub-sequences
  is_any_subsequence_safe <- function(sequence) {
    # Iterate through each element in the sequence
    for (index in seq_along(sequence)) {
      # Exclude the current element to create a sub-sequence
      subsequence <- sequence[-index]
      
      # Compute differences for the sub-sequence
      differences <- diff(subsequence)
      
      # Check if the sub-sequence satisfies the "safe" condition
      if (all(differences > 0 & differences < 4) || all(differences < 0 & differences > -4)) {
        return(TRUE)
      }
    }
    return(FALSE)
  }
  
  # Apply the "safe" condition to each sequence and sum up the results
  safe_count <- sum(sapply(sequences, is_any_subsequence_safe))
  
  return(safe_count)
}

# Call the functions
safe_sequence_result <- safe_sequence_count()
safe_subsequence_result <- safe_subsequence_count()

# Print results
print(safe_sequence_result)
print(safe_subsequence_result)
